ports/rp2/machine_uart.c

```diff
@@ -270,7 +303,8 @@ STATIC void machine_uart_init_helper(machine_uart_obj_t *self, size_t n_args, co
         self->flow = args[ARG_flow].u_int;
     }
 
-    self->read_lock = false;
+    read_mutex_init(self);
+    write_mutex_init(self);
```

Comment:
Is it correct to init these every time the constructor/`init()` is called?  What if IRQs are called while these are being re-init'd?

Maybe it's possible to do the following instead of calling these init functions, replacing the definitions of the mutex's above:
```c
auto_init_mutex(write_mutex_0);
auto_init_mutex(write_mutex_1);
auto_init_mutex(read_mutex_0);
auto_init_mutex(read_mutex_1);
```

Reply:
At the time when the mutexes are initialised the ISR's have not been set. This is done [here](https://github.com/micropython/micropython/blob/fe8b47e29fb40566fa35e97e02ab06cfab19bf6e/ports/rp2/machine_uart.c#L338).

Reply:
But what about this case:
```python
u = UART(0, 9600)
u.write(b'data')
u.init(115200)
```
In the call to `.init(115200)` there could be IRQs being processed during the init.

Reply:
@peterhinch any comment on the above, regarding re-init'ing?

Reply:
Sorry, I missed that comment. I take your point and that probably is the simplest fix. I'll re-test and push an update.

Reply:
Now done. Running happily here at >100K messages.